CN Testing Page Quiz Functionality

chrisn02 opened the pull request:
The user is now able to start and complete a quiz that has been assigned to them. It will be displayed in the "Tests to Complete" section and once clicked the quiz will begin. 
![image](https://github.com/Jforjo/CSD/assets/59711723/5d405b95-4540-4030-bcdf-1f77dc50f0a2)
![image](https://github.com/Jforjo/CSD/assets/59711723/20009dec-e751-4098-913e-91a7105bfa9e) The questions will be loaded. The correct answer is stored server side rather than client side to prevent users from using inspect element to find the correct answer
![image](https://github.com/Jforjo/CSD/assets/59711723/5c9ee4f2-f4e2-4fe3-a577-2198a764a46f) When the user selects an answer a button will appear to go to the next question.

![image](https://github.com/Jforjo/CSD/assets/59711723/e5511b33-dba3-4e69-9f1e-1df49420d3bc) The user has now advanced to the next question. There is also a progress bar that goes up with each question

![image](https://github.com/Jforjo/CSD/assets/59711723/1fe57818-e262-4872-8af4-eaee2b223cf7) If there is a question with less than 4 answers, then only the needed number of boxes will appear

![image](https://github.com/Jforjo/CSD/assets/59711723/3e7d974d-556c-4f8c-9ac2-de9bcf533d87) Once the user has completed all questions a popup will appear saying that quiz is complete and there will be a button to view the results

![image](https://github.com/Jforjo/CSD/assets/59711723/67ca551c-0651-47e0-9037-0f24e8524a23) Another popup will appear with the results and these results will be saved to the database

![image](https://github.com/Jforjo/CSD/assets/59711723/047e6f1c-5205-4afb-bde2-4f07fd98492b) The quiz that was just completed has been moved to the "Completed Tests" section

chrisn02 committed:
Loads subject name and quiz name

Also moved testing-page.php to pages/student

chrisn02 committed:
Now loads questions and asnwers from each quiz

Loads questions from each quiz using GetQuizSet procedure

chrisn02 committed:
The user can now load the next question after clicking an answer

The user can click on an answer, which would cause a button to appear, which can be clicked to load the next question. Also stores the number of questions and the current question number which are then displayed

chrisn02 committed:
Stores users answer and progress bar works

Progress bar works by setting it to 0% on question 1, then incrementing based off current question number

Cannot currently store correct answer as it's broken

chrisn02 committed:
Added check to see if answer box is blank

Also added some styling to next question button and added message after quiz has been completed (currently as a placeholder)

chrisn02 committed:
Fixed a bug to do with the user's answer choice

The user's answer is supposed to be 1,2,3 or 4, depending on which answer they click which it correctly does for question 1, but for question 2 it sets their answer to either 5,6,7 or 8 which it shouldn't do. This was fixed by changing the answer selection to be within the question container. Still need to make it so it only sets the user's choice after clicking the next question button.

chrisn02 committed:
Gets correct answer for each question and displays results

The program now gets the correct answer for each question and after all questions have been answered then the results are displayed which includes the score, number of correct questions and percentage. Needs proper styling though

chrisn02 committed:
Added some styling to modal

User is also redirected to test selection page after clicking the finish button on the results modal

chrisn02 committed:
More styling to modal

chrisn02 committed:
After user completes quiz, it now saves it to database

Saves the points, correct answers, date completed and sets the "completed" field from 0 to 1, which moves it to the completes tests section

chrisn02 committed:
Correct answer is no longer stored client side

Moved the correct answer validation from client side to server side as the user could use inspect element beforehand to find the correct answer. Also added some functions to better organise the code. Also when the user selects an answer it stays highlighted so they know which one they clicked

Jforjo commented on pages/student/testing-page.php:
```diff
@@ -0,0 +1,232 @@
+    //Get question number and total number of questions
+    $questionNumber = 0;
+    $totalQuestions = count($allQuestions);
+}
```
I would heavily suggest using `GetQuizSet` to get a set a questions, as this randomises the order and allows you to limit how many questions are in each quiz. However, what you have now is fine.

chrisn02 replied:
I originally did use GetQuizSet, but I was having an error with it the other day so I switched to GetAllQuestions. Have switched back and it now works

Jforjo commented on pages/student/testing-page.php:
```diff
@@ -0,0 +1,232 @@
+        <?php foreach ($allQuestions as $question): ?>
+        <!-- Increment the question number -->
+        <?php $questionNumber++; ?>
+        <?php $correctAnswer = intval($question['correctAnswer']);?>
```
I don't see the purpose of storing the correct answer here.

chrisn02 replied:
That was a leftover from when I was storing the correct answer client-side

Jforjo commented on pages/student/testing-page.php:
```diff
@@ -0,0 +1,232 @@
+        <p id="total-score"><i class="fas fa-trophy"></i>Points Earned: </p>
+        <p id="correct-answers"><i class="fas fa-check"></i>Correct questions: </p>
+        <p id="percentage"><i class="fas fa-chart-line"></i>Percentage: </p>
+        <button onclick="window.location.href='index.php'">Finish</button>
```
Just use an `<a>` tag.

Jforjo commented on pages/student/testing-page.php:
```diff
@@ -0,0 +1,232 @@
+        if (this.readyState == 4 && this.status == 200) {
+            // The server has responded
+            var response = JSON.parse(this.responseText);
+            console.log(this.responseText)
```
I think this was for testing purposes, so it can probably be removed now.

Jforjo commented on php/checkAnswer.php:
```diff
@@ -0,0 +1,25 @@
+$conn = newConn();
+
+// Get the correct answer from the database
+$stmt = $conn->prepare("SELECT correctAnswer FROM questions WHERE questionID = :questionID");
```
Make a prepared statement for this.

chrisn02 replied:
Could you clarify this please, I think it's already in a prepared statement but correct me if wrong

Jforjo replied:
I meant a prepared procedure/function.
Your SQL should look like "CALL myFunction();" and not what you have now.

chrisn02 replied:
Ah, got it. Will do that now

Jforjo commented on php/updateCompletedQuiz.php:
```diff
@@ -0,0 +1,27 @@
+        UPDATE `studentQuizLink`
+        SET `completed` = :completed, `questionCount` = :questionCount, `correctCount` = :correctCount, `points` = :points, `dateCompleted` = :dateCompleted
+        WHERE `studentQuizLinkID` = :studentQuizLinkID
+    ";
```
Make a prepared statement for this.

Jforjo reviewed (changes requested):
I think you should also start adding checks to your files - only let students access the index, testing-page, checkanswer, updatecompletedquiz and any other that only a student needs. Also, redirect them back to the login screen if they are not logged in.
See comments for more.

chrisn02 committed:
Have made requested changes

Added checks to see if the user is a student and if they aren't then they will be given an error message. Have added a check to see if the user is logged in and if they aren't then they will be redirected to the login page. Have also replaced the SQL queries with procedures

chrisn02 commented:
Have now added checks to only allow the students to access the necessary pages and a check to redirect the user to the login page if they are not logged in. Also used a procedure for the SQL queries

Jforjo commented on pages/student/index.php:
```diff
@@ -20,11 +20,25 @@
     </header>
     <?php 
     require_once('../../php/connection.php');
+    require_once(__DIR__ . '/../../php/dbfuncs.php');
```
The connection file is required in the dbfuncs file, so you only need the dbfuncs require line.

chrisn02 replied:
Sorted

chrisn02 committed:
Removed require connection.php line

gchodancar19 reviewed (approved):
I think the fields created looks successful as it shows when the next field is selected and also shows the progress of when of how much question have been completed and at last shows the test results.
